File: ios-6/Tracqer/Tracqer/APIClient.m
#import "APIClient.h"
#import "Crypto.h"
#import "VinylRecord.h"

NSString * const TracqerAPIErrorDomain = @"TracqerAPI";

static APIClient *gSharedClient = nil;

static NSError *APIError(NSInteger code, NSString *msg) {
    return [NSError errorWithDomain:TracqerAPIErrorDomain
                               code:code
                           userInfo:@{NSLocalizedDescriptionKey: msg ?: @"Unknown error"}];
}

static NSString *NormaliseBaseURL(NSString *u) {
    if ([u hasSuffix:@"/"]) return [u substringToIndex:u.length - 1];
    return u;
}

static NSString *URLQueryEscape(NSString *s) {
    CFStringRef esc = CFURLCreateStringByAddingPercentEscapes(NULL, (__bridge CFStringRef)s, NULL,
                                                              CFSTR(":/?#[]@!$&'()*+,;="),
                                                              kCFStringEncodingUTF8);
    return CFBridgingRelease(esc);
}

#pragma mark - Internal connection wrapper (accepts any server trust)

// NSURLConnection delegate-based wrapper that accepts any TLS cert. Used because
// (a) sendAsynchronousRequest: cannot handle auth challenges, and (b) we talk
// only to the user's own server — and on the Mountain Lion build host the
// system root CA store is too old to validate modern Let's Encrypt certs.
typedef void (^TracqerHTTPCompletion)(NSURLResponse *response, NSData *data, NSError *error);

@interface _TracqerHTTPConnection : NSObject <NSURLConnectionDataDelegate> {
    NSMutableData *_buffer;
    NSURLResponse *_response;
    TracqerHTTPCompletion _completion;
    NSURLConnection *_connection;
    _TracqerHTTPConnection *_selfRetain;
}
+ (void)performRequest:(NSURLRequest *)req completion:(TracqerHTTPCompletion)completion;
@end

@implementation _TracqerHTTPConnection

+ (void)performRequest:(NSURLRequest *)req completion:(TracqerHTTPCompletion)completion {
    _TracqerHTTPConnection *c = [[self alloc] init];
    c->_buffer     = [NSMutableData data];
    c->_completion = [completion copy];
    c->_selfRetain = c;  // keep alive until connection finishes
    c->_connection = [[NSURLConnection alloc] initWithRequest:req delegate:c startImmediately:NO];
    [c->_connection scheduleInRunLoop:[NSRunLoop mainRunLoop] forMode:NSDefaultRunLoopMode];
    [c->_connection start];
}

- (void)connection:(NSURLConnection *)connection willSendRequestForAuthenticationChallenge:(NSURLAuthenticationChallenge *)challenge {
    if ([challenge.protectionSpace.authenticationMethod isEqualToString:NSURLAuthenticationMethodServerTrust]) {
        NSURLCredential *cred = [NSURLCredential credentialForTrust:challenge.protectionSpace.serverTrust];
        [challenge.sender useCredential:cred forAuthenticationChallenge:challenge];
    } else {
        [challenge.sender continueWithoutCredentialForAuthenticationChallenge:challenge];
    }
}

- (void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLResponse *)response {
    _response = response;
    _buffer.length = 0;
}

- (void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data {
    [_buffer appendData:data];
}

- (void)connectionDidFinishLoading:(NSURLConnection *)connection {
    if (_completion) _completion(_response, [_buffer copy], nil);
    _selfRetain = nil;
}

- (void)connection:(NSURLConnection *)connection didFailWithError:(NSError *)error {
    if (_completion) _completion(nil, nil, error);
    _selfRetain = nil;
}

@end

#pragma mark -

@implementation APIClient

+ (instancetype)sharedClient {
    return gSharedClient;
}

+ (void)configureSharedWithBaseURL:(NSString *)baseURL key:(NSData *)key token:(NSString *)token {
    gSharedClient = [[APIClient alloc] initWithBaseURL:baseURL key:key token:token];
}

+ (void)resetShared {
    gSharedClient = nil;
}

- (instancetype)initWithBaseURL:(NSString *)baseURL key:(NSData *)key token:(NSString *)token {
    if ((self = [super init])) {
        _baseURL = [NormaliseBaseURL(baseURL) copy];
        _key     = [key copy];
        _token   = [token copy];
    }
    return self;
}

#pragma mark - Static endpoints (ping, verify)

+ (void)pingBaseURL:(NSString *)baseURL completion:(APIVoidCompletion)completion {
    NSString *full = [NormaliseBaseURL(baseURL) stringByAppendingString:@"/ping"];
    NSURL *url = [NSURL URLWithString:full];
    if (!url) { completion(NO, APIError(1, @"Invalid server URL")); return; }

    NSURLRequest *req = [NSURLRequest requestWithURL:url
                                         cachePolicy:NSURLRequestReloadIgnoringLocalCacheData
                                     timeoutInterval:10];
    [_TracqerHTTPConnection performRequest:req completion:^(NSURLResponse *resp, NSData *data, NSError *err) {
        if (err) { completion(NO, err); return; }
        NSInteger code = [(NSHTTPURLResponse *)resp statusCode];
        if (code == 200) { completion(YES, nil); return; }
        completion(NO, APIError(code, [NSString stringWithFormat:@"Ping returned HTTP %ld", (long)code]));
    }];
}

+ (void)verifyBaseURL:(NSString *)baseURL key:(NSData *)key token:(NSString *)token completion:(APIVoidCompletion)completion {
    NSError *encErr = nil;
    NSDictionary *envelope = [Crypto encryptObject:@{} withKey:key error:&encErr];
    if (!envelope) { completion(NO, encErr ?: APIError(2, @"Encrypt failed")); return; }

    NSData *body = [NSJSONSerialization dataWithJSONObject:envelope options:0 error:NULL];
    NSString *full = [NormaliseBaseURL(baseURL) stringByAppendingString:@"/api/v1/auth/verify"];
    NSMutableURLRequest *req = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:full]
                                                       cachePolicy:NSURLRequestReloadIgnoringLocalCacheData
                                                   timeoutInterval:15];
    req.HTTPMethod = @"POST";
    req.HTTPBody = body;
    [req setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    [req setValue:[NSString stringWithFormat:@"Bearer %@", token] forHTTPHeaderField:@"Authorization"];

    [_TracqerHTTPConnection performRequest:req completion:^(NSURLResponse *resp, NSData *data, NSError *err) {
        if (err) { completion(NO, err); return; }
        NSInteger code = [(NSHTTPURLResponse *)resp statusCode];
        if (code != 200) {
            completion(NO, APIError(code, [NSString stringWithFormat:@"Verify HTTP %ld", (long)code]));
            return;
        }
        id env = [NSJSONSerialization JSONObjectWithData:data options:0 error:NULL];
        if (![env isKindOfClass:[NSDictionary class]]) { completion(NO, APIError(3, @"Bad verify response")); return; }
        NSError *dErr = nil;
        id decrypted = [Crypto decryptEnvelope:env withKey:key error:&dErr];
        if (![decrypted isKindOfClass:[NSDictionary class]]) { completion(NO, dErr ?: APIError(4, @"Decrypt failed")); return; }
        BOOL valid = [((NSDictionary *)decrypted)[@"valid"] boolValue];
        if (valid) completion(YES, nil);
        else completion(NO, APIError(5, @"Wrong password"));
    }];
}

#pragma mark - Encrypted request core

- (void)performMethod:(NSString *)method
                 path:(NSString *)path
                 body:(id)bodyObject
           expectJSON:(BOOL)expectJSON
           completion:(void (^)(id decodedJSON, NSError *error))completion {

    NSString *full = [self.baseURL stringByAppendingString:path];
    NSURL *url = [NSURL URLWithString:full];
    if (!url) { completion(nil, APIError(1, @"Bad URL")); return; }

    NSMutableURLRequest *req = [NSMutableURLRequest requestWithURL:url
                                                       cachePolicy:NSURLRequestReloadIgnoringLocalCacheData
                                                   timeoutInterval:20];
    req.HTTPMethod = method;
    [req setValue:[NSString stringWithFormat:@"Bearer %@", self.token] forHTTPHeaderField:@"Authorization"];

    if (bodyObject) {
        NSError *encErr = nil;
        NSDictionary *env = [Crypto encryptObject:bodyObject withKey:self.key error:&encErr];
        if (!env) { completion(nil, encErr); return; }
        req.HTTPBody = [NSJSONSerialization dataWithJSONObject:env options:0 error:NULL];
        [req setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    }

    [_TracqerHTTPConnection performRequest:req completion:^(NSURLResponse *resp, NSData *data, NSError *err) {
        if (err) { completion(nil, err); return; }
        NSInteger code = [(NSHTTPURLResponse *)resp statusCode];
        if (code >= 300) {
            NSString *msg = data ? [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding] : @"";
            completion(nil, APIError(code, [NSString stringWithFormat:@"HTTP %ld: %@", (long)code, msg]));
            return;
        }
        if (!expectJSON || code == 204 || data.length == 0) {
            completion(nil, nil); return;
        }
        id env = [NSJSONSerialization JSONObjectWithData:data options:0 error:NULL];
        if (![env isKindOfClass:[NSDictionary class]]) { completion(nil, APIError(3, @"Bad response envelope")); return; }
        NSError *dErr = nil;
        id decrypted = [Crypto decryptEnvelope:env withKey:self.key error:&dErr];
        if (!decrypted) { completion(nil, dErr ?: APIError(4, @"Decrypt failed")); return; }
        completion(decrypted, nil);
    }];
}

#pragma mark - Records CRUD

- (void)listRecordsSearch:(NSString *)search page:(NSInteger)page limit:(NSInteger)limit completion:(APIRecordListCompletion)completion {
    NSMutableString *path = [NSMutableString stringWithFormat:@"/api/v1/records?page=%ld&limit=%ld", (long)page, (long)limit];
    if (search.length > 0) [path appendFormat:@"&search=%@", URLQueryEscape(search)];

    [self performMethod:@"GET" path:path body:nil expectJSON:YES completion:^(id json, NSError *err) {
        if (err) { completion(nil, 0, err); return; }
        NSArray *raw = json[@"records"];
        NSMutableArray *records = [NSMutableArray arrayWithCapacity:raw.count];
        for (NSDictionary *d in raw) [records addObject:[VinylRecord recordFromDictionary:d]];
        NSInteger total = [json[@"total"] integerValue];
        completion(records, total, nil);
    }];
}

- (void)getRecord:(NSString *)recordId completion:(APIRecordCompletion)completion {
    [self performMethod:@"GET" path:[NSString stringWithFormat:@"/api/v1/records/%@", recordId] body:nil expectJSON:YES completion:^(id json, NSError *err) {
        if (err) { completion(nil, err); return; }
        completion([VinylRecord recordFromDictionary:json], nil);
    }];
}

- (void)createRecord:(VinylRecord *)record completion:(APIRecordCompletion)completion {
    [self performMethod:@"POST" path:@"/api/v1/records" body:[record toDictionaryForInput] expectJSON:YES completion:^(id json, NSError *err) {
        if (err) { completion(nil, err); return; }
        completion([VinylRecord recordFromDictionary:json], nil);
    }];
}

- (void)updateRecord:(VinylRecord *)record completion:(APIRecordCompletion)completion {
    NSString *path = [NSString stringWithFormat:@"/api/v1/records/%@", record.recordId];
    [self performMethod:@"PATCH" path:path body:[record toDictionaryForInput] expectJSON:YES completion:^(id json, NSError *err) {
        if (err) { completion(nil, err); return; }
        completion([VinylRecord recordFromDictionary:json], nil);
    }];
}

- (void)deleteRecord:(NSString *)recordId completion:(APIVoidCompletion)completion {
    NSString *path = [NSString stringWithFormat:@"/api/v1/records/%@", recordId];
    [self performMethod:@"DELETE" path:path body:nil expectJSON:NO completion:^(id json, NSError *err) {
        completion(err == nil, err);
    }];
}

#pragma mark - Photos (read-only)

- (NSURL *)photoURLForRecord:(NSString *)recordId photo:(Photo *)photo size:(NSInteger)size {
    NSString *path;
    if ([photo isDiscPhoto] && photo.discNumber) {
        path = [NSString stringWithFormat:@"/api/v1/records/%@/photos/%@/%@", recordId, photo.photoType, photo.discNumber];
    } else {
        path = [NSString stringWithFormat:@"/api/v1/records/%@/photos/%@", recordId, photo.photoType];
    }
    NSString *full = [NSString stringWithFormat:@"%@%@?token=%@&size=%ld",
                      self.baseURL, path, self.token, (long)size];
    return [NSURL URLWithString:full];
}

- (void)fetchPhotoData:(NSURL *)url completion:(APIDataCompletion)completion {
    static NSCache *cache;
    static dispatch_once_t once;
    dispatch_once(&once, ^{
        cache = [[NSCache alloc] init];
        cache.totalCostLimit = 16 * 1024 * 1024;  // ~16 MB in-memory image cache
    });

    NSString *key = url.absoluteString;
    NSData *cached = [cache objectForKey:key];
    if (cached) { completion(cached, nil); return; }

    NSURLRequest *req = [NSURLRequest requestWithURL:url
                                         cachePolicy:NSURLRequestReturnCacheDataElseLoad
                                     timeoutInterval:30];
    [_TracqerHTTPConnection performRequest:req completion:^(NSURLResponse *resp, NSData *data, NSError *err) {
        if (err) { completion(nil, err); return; }
        NSInteger code = [(NSHTTPURLResponse *)resp statusCode];
        if (code != 200) { completion(nil, APIError(code, [NSString stringWithFormat:@"Photo HTTP %ld", (long)code])); return; }
        if (data) [cache setObject:data forKey:key cost:data.length];
        completion(data, nil);
    }];
}

@end

File: ios/VinylCollection/APIClient.swift
import Foundation
import UIKit

// MARK: - Errors

enum APIError: LocalizedError {
    case http(Int, String)
    case noData

    var errorDescription: String? {
        switch self {
        case .http(let code, let msg): return "Server error \(code): \(msg)"
        case .noData:                  return "No data received"
        }
    }
}

// MARK: - APIClient

final class APIClient: NSObject {

    let baseURL: String
    let key: Data
    let token: String

    private lazy var session: URLSession = {
        URLSession(configuration: .default, delegate: self, delegateQueue: nil)
    }()

    init(baseURL: String, key: Data, token: String) {
        self.baseURL = baseURL.hasSuffix("/") ? String(baseURL.dropLast()) : baseURL
        self.key     = key
        self.token   = token
    }

    // MARK: - Core request

    func request<T: Decodable>(_ method: String, path: String, body: (any Encodable)? = nil) async throws -> T {
        var req = URLRequest(url: URL(string: baseURL + path)!)
        req.httpMethod = method
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        if let body {
            let dict     = try body.toDictionary()
            let envelope = try CryptoService.encrypt(key: key, object: dict)
            req.httpBody = try JSONSerialization.data(withJSONObject: envelope)
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }

        return try await performAndDecrypt(req)
    }

    // No-body request returning nothing (DELETE → 204)
    func requestNoContent(_ method: String, path: String) async throws {
        var req = URLRequest(url: URL(string: baseURL + path)!)
        req.httpMethod = method
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        let (_, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse, http.statusCode < 300 else {
            throw APIError.http((response as? HTTPURLResponse)?.statusCode ?? 0, "")
        }
    }

    // Multipart upload for photos
    func uploadPhoto(path: String, imageData: Data, mimeType: String) async throws {
        var req = URLRequest(url: URL(string: baseURL + path)!)
        req.httpMethod = "POST"
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let boundary = UUID().uuidString
        req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"photo.jpg\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: \(mimeType)\r\n\r\n".data(using: .utf8)!)
        body.append(imageData)
        body.append("\r\n--\(boundary)--\r\n".data(using: .utf8)!)
        req.httpBody = body

        let (_, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse, http.statusCode < 300 else {
            throw APIError.http((response as? HTTPURLResponse)?.statusCode ?? 0, "Upload failed")
        }
    }

    // Multipart create-with-photos
    func createRecord(input: RecordInput, photos: [String: Data]) async throws -> VinylRecord {
        var req = URLRequest(url: URL(string: baseURL + "/api/v1/records/upload")!)
        req.httpMethod = "POST"
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let boundary = UUID().uuidString
        req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        let dict     = try input.toDictionary()
        let envelope = try CryptoService.encrypt(key: key, object: dict)
        let metaJSON = try JSONSerialization.data(withJSONObject: envelope)

        var body = Data()
        body.append("--\(boundary)\r\nContent-Disposition: form-data; name=\"metadata\"\r\n\r\n".data(using: .utf8)!)
        body.append(metaJSON)
        body.append("\r\n".data(using: .utf8)!)

        for (fieldName, imgData) in photos {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"\(fieldName)\"; filename=\"\(fieldName).jpg\"\r\n".data(using: .utf8)!)
            body.append("Content-Type: image/jpeg\r\n\r\n".data(using: .utf8)!)
            body.append(imgData)
            body.append("\r\n".data(using: .utf8)!)
        }
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)
        req.httpBody = body

        return try await performAndDecrypt(req)
    }

    // MARK: - Photo URL

    func photoURL(recordId: String, photoType: String, discNumber: Int? = nil, size: String = "640") -> URL {
        let isDisc = photoType == "disc_front" || photoType == "disc_back"
        let path: String
        if isDisc, let disc = discNumber {
            path = "/api/v1/records/\(recordId)/photos/\(photoType)/\(disc)"
        } else {
            path = "/api/v1/records/\(recordId)/photos/\(photoType)"
        }
        return URL(string: "\(baseURL)\(path)?token=\(token)&size=\(size)")!
    }

    // MARK: - Image fetch (self-signed cert aware)

    func fetchImage(url: URL) async throws -> UIImage {
        let req = URLRequest(url: url)
        let (data, _) = try await session.data(for: req)
        guard let img = UIImage(data: data) else { throw APIError.noData }
        return img
    }

    // MARK: - Helpers

    private func performAndDecrypt<T: Decodable>(_ req: URLRequest) async throws -> T {
        let (data, response) = try await session.data(for: req)
        if let http = response as? HTTPURLResponse, http.statusCode >= 300 {
            let msg = String(data: data, encoding: .utf8) ?? ""
            throw APIError.http(http.statusCode, msg)
        }
        let envelope = try JSONSerialization.jsonObject(with: data) as! [String: Any]
        let decrypted = try CryptoService.decrypt(key: key, envelope: envelope)
        let plain = try JSONSerialization.data(withJSONObject: decrypted)
        let decoder = JSONDecoder()
        return try decoder.decode(T.self, from: plain)
    }
}

// MARK: - Self-signed certificate support

extension APIClient: URLSessionDelegate {
    func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge,
                    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        guard challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
              let trust = challenge.protectionSpace.serverTrust else {
            completionHandler(.cancelAuthenticationChallenge, nil)
            return
        }
        // Accept any certificate — this app only talks to our home server
        completionHandler(.useCredential, URLCredential(trust: trust))
    }
}

// MARK: - Encodable → Dictionary helper

extension Encodable {
    func toDictionary() throws -> [String: Any] {
        let data = try JSONEncoder().encode(self)
        return try JSONSerialization.jsonObject(with: data) as! [String: Any]
    }
}

// MARK: - Convenience record API methods

extension APIClient {

    func listRecords(search: String? = nil, page: Int = 1, limit: Int = 100) async throws -> RecordsResponse {
        var params: [String] = ["page=\(page)", "limit=\(limit)"]
        if let s = search, !s.isEmpty { params.append("search=\(s.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? s)") }
        return try await request("GET", path: "/api/v1/records?\(params.joined(separator: "&"))")
    }

    func getRecord(_ id: String) async throws -> VinylRecord {
        try await request("GET", path: "/api/v1/records/\(id)")
    }

    func updateRecord(_ id: String, input: RecordInput) async throws -> VinylRecord {
        try await request("PATCH", path: "/api/v1/records/\(id)", body: input)
    }

    func deleteRecord(_ id: String) async throws {
        try await requestNoContent("DELETE", path: "/api/v1/records/\(id)")
    }

    func deletePhoto(recordId: String, photoType: String, discNumber: Int? = nil) async throws {
        let isDisc = photoType == "disc_front" || photoType == "disc_back"
        let path: String
        if isDisc, let disc = discNumber {
            path = "/api/v1/records/\(recordId)/photos/\(photoType)/\(disc)"
        } else {
            path = "/api/v1/records/\(recordId)/photos/\(photoType)"
        }
        try await requestNoContent("DELETE", path: path)
    }

    func uploadSleevePhoto(recordId: String, photoType: String, imageData: Data) async throws {
        try await uploadPhoto(path: "/api/v1/records/\(recordId)/photos/\(photoType)", imageData: imageData, mimeType: "image/jpeg")
    }

    func uploadDiscPhoto(recordId: String, photoType: String, discNumber: Int, imageData: Data) async throws {
        try await uploadPhoto(path: "/api/v1/records/\(recordId)/photos/\(photoType)/\(discNumber)", imageData: imageData, mimeType: "image/jpeg")
    }

    static func ping(baseURL: String) async -> Bool {
        guard let url = URL(string: "\(baseURL)/ping") else { return false }
        do {
            // Temporary session that also accepts self-signed certs
            let delegate = PingDelegate()
            let session = URLSession(configuration: .default, delegate: delegate, delegateQueue: nil)
            let (_, resp) = try await session.data(from: url)
            return (resp as? HTTPURLResponse)?.statusCode == 200
        } catch { return false }
    }

    static func verify(baseURL: String, key: Data, token: String) async -> Bool {
        guard let url = URL(string: "\(baseURL)/api/v1/auth/verify") else { return false }
        do {
            let envelope = try CryptoService.encrypt(key: key, object: [:])
            let body     = try JSONSerialization.data(withJSONObject: envelope)
            var req      = URLRequest(url: url)
            req.httpMethod = "POST"
            req.httpBody   = body
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

            let delegate = PingDelegate()
            let session  = URLSession(configuration: .default, delegate: delegate, delegateQueue: nil)
            let (data, resp) = try await session.data(for: req)
            guard (resp as? HTTPURLResponse)?.statusCode == 200 else { return false }
            let envResp = try JSONSerialization.jsonObject(with: data) as! [String: Any]
            let decrypted = try CryptoService.decrypt(key: key, envelope: envResp) as? [String: Any]
            return decrypted?["valid"] as? Bool == true
        } catch { return false }
    }
}

private final class PingDelegate: NSObject, URLSessionDelegate {
    func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge,
                    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        guard let trust = challenge.protectionSpace.serverTrust else {
            completionHandler(.cancelAuthenticationChallenge, nil); return
        }
        completionHandler(.useCredential, URLCredential(trust: trust))
    }
}
